### PR TITLE
Make viaMat on identity Flow preserve upstreams/downstreams #22585

### LINKED ISSUE #22585: Flow.map(???).mapConcat(???) may deadlock
Deadlock occurs depending on whether stream fusing is turned on and "surrounding context" (for the lack of a better term).
Here's the code demonstrating the problem:
```scala
object StreamDeadlock extends App {
  import akka.actor._
  import akka.stream._
  import akka.stream.scaladsl._
  import scala.concurrent.Await

  implicit val system = ActorSystem()

  val noFusingMaterializer = ActorMaterializer(ActorMaterializerSettings(system).withAutoFusing(false))
  val fusingMaterializer = ActorMaterializer(ActorMaterializerSettings(system).withAutoFusing(true))

  val source = Source(1 to 20)
  val thisCanDeadlockSometimes = Flow[Int].map(_ + 1).mapConcat(x => List(x))
  val printFlow = Flow[Int].map(println)
  val sink = Flow[Int].via(thisCanDeadlockSometimes).via(printFlow).toMat(Sink.ignore)(Keep.right)

  println("Running inlined stream without fusing: OK")
  Await.result(source.via(thisCanDeadlockSometimes).via(printFlow).runWith(Sink.ignore)(noFusingMaterializer), Duration.Inf)

  println("Running composite stream with fusing: OK")
  Await.result(source.runWith(sink)(fusingMaterializer), Duration.Inf)

  println("Running composite stream without fusing: DEADLOCK")
  Await.result(source.runWith(sink)(noFusingMaterializer), Duration.Inf)

  system.terminate()
}
```

Expected output:
```
Running inlined stream without fusing: OK
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
Running composite stream with fusing: OK
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
Running composite stream without fusing: DEADLOCK
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
```

Actual output:
```
Running inlined stream without fusing: OK
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
Running composite stream with fusing: OK
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
Running composite stream without fusing: DEADLOCK
(wait indefinitely)
```
Tested on akka 2.4.17 / Scala 2.12.1

### PR DESCRIPTION
**Pull request by bantonsson**
Fixes #22585

Passes all `akka-stream-tests`, `akka-stream-test-tck` and also `akka-http` tests.

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala [hunk 1]
```diff
@@ -593,8 +593,21 @@ class FlowSpec extends StreamSpec(ConfigFactory.parseString("akka.actor.debug.re
       import Attributes._
       val f: Flow[Int, Int, NotUsed] = Flow[Int].map(_ + 1).async.addAttributes(none).named("name")
 
-      f.module.attributes.getFirst[Name] shouldEqual Some(Name("name"))
-      f.module.attributes.getFirst[Attributes.AsyncBoundary.type] shouldEqual Some(AsyncBoundary)
+      f.module.attributes.get[Name] shouldEqual Some(Name("name"))
+      f.module.attributes.get[Attributes.AsyncBoundary.type] shouldEqual Some(AsyncBoundary)
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala]: Was these changes a consequence of the fix? (least specific vs most specific attribute)

**Review by johanandren — APPROVED** [review]
LGTM with one question. Great detective work @bantonsson !

**Review by patriknw — APPROVED** [review]
LGTM given that tests are passing (thanks for running http also)
Great detective work!

**Review by ktoso — APPROVED** [review]
Wow what a catch! Thanks a ton

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala (hunk 1 above)
Comment (bantonsson) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala]: Yes, the test looked up the least specific `Name` which actually is `map` for a `map` operation, when Attributes are preserved, while the most specific is `name`.

**Pull request merged**